src/txn/txn_ckpt.c

```diff
@@ -532,26 +531,44 @@ __wt_txn_checkpoint(WT_SESSION_IMPL *session, const char *cfg[])
 	WT_ERR(__wt_txn_commit(session, NULL));
 
 	/*
-	 * If any tree was dirty, we will have updated the metadata with the
-	 * new checkpoint information.  If the metadata is clean, all other
-	 * trees must have been clean.
-	 *
-	 * Disable metadata tracking during the metadata checkpoint.
-	 *
-	 * We don't lock old checkpoints in the metadata file: there is no way
-	 * to open one.  We are holding other handle locks, it is not safe to
-	 * lock conn->spinlock.
+	 * Checkpoint the metadata file if this is a system wide checkpoint or
+	 * logging is disabled. Otherwise ensure that the log file has been
+	 * flushed so that the checkpoint is stable on disk.
+	 * It is not OK to checkpoint the metadata if logging is enabled and
+	 * this isn't a full checkpoint since checkpointing the metadata
+	 * updates the start LSN for recovery.
 	 */
-	session->isolation = txn->isolation = WT_ISO_READ_UNCOMMITTED;
-	saved_meta_next = session->meta_track_next;
-	session->meta_track_next = NULL;
-	WT_WITH_DHANDLE(session,
-	    session->meta_dhandle, ret = __wt_checkpoint(session, cfg));
-	session->meta_track_next = saved_meta_next;
-	WT_ERR(ret);
-
-	WT_ERR(__checkpoint_verbose_track(session,
-	    "metadata sync completed", &verb_timer));
+	if (full || !logging) {
+		/*
+		 * If any tree was dirty, we will have updated the metadata
+		 * with the new checkpoint information.  If the metadata is
+		 * clean, all other trees must have been clean.
+		 *
+		 * Disable metadata tracking during the metadata checkpoint.
+		 *
+		 * We don't lock old checkpoints in the metadata file: there is
+		 * no way to open one.  We are holding other handle locks, it
+		 * is not safe to lock conn->spinlock.
+		 */
+		session->isolation = txn->isolation = WT_ISO_READ_UNCOMMITTED;
+		saved_meta_next = session->meta_track_next;
+		session->meta_track_next = NULL;
+		WT_WITH_DHANDLE(session,
+		    session->meta_dhandle, ret = __wt_checkpoint(session, cfg));
+		session->meta_track_next = saved_meta_next;
+		WT_ERR(ret);
+
+		WT_ERR(__checkpoint_verbose_track(session,
+		    "metadata sync completed", &verb_timer));
+	} else if (logging)
```

Comment:
fwiw, I think the condition here is redundant.  That is, `logging` will always be true in the else case.